File: AllegroDotNet.Wrapper/src/AlJoystick.cpp
#include "PCH.h"

#include "Al.h"
#include "AllegroEventSource.h"
#include "AllegroJoystick.h"
#include "AllegroJoystickState.h"
#include "JoyFlags.h"

namespace AllegroDotNet::Wrapper
{
	AllegroEventSource^ Al::GetJoystickEventSource()
	{
		auto* nativeEventSource = al_get_joystick_event_source();
		return gcnew AllegroEventSource(nativeEventSource);
	}

	AllegroJoystick^ Al::GetJoystick(Int32 num)
	{
		auto* nativeReturnValue = al_get_joystick(num);
		return gcnew AllegroJoystick(nativeReturnValue);
	}

	Boolean Al::GetJoystickActive(AllegroJoystick^ joy)
	{
		auto* nativeJoy = joy->GetNativePointer();
		return al_get_joystick_active(nativeJoy);
	}

	Boolean Al::InstallJoystick()
	{
		return al_install_joystick();
	}

	Boolean Al::IsJoystickInstalled()
	{
		return al_is_joystick_installed();
	}

	Boolean Al::ReconfigureJoysticks()
	{
		return al_reconfigure_joysticks();
	}

	Int32 Al::GetJoystickNumAxes(AllegroJoystick^ joy, Int32 stick)
	{
		auto* nativeJoy = joy->GetNativePointer();
		return al_get_joystick_num_axes(nativeJoy, stick);
	}

	Int32 Al::GetJoystickNumButtons(AllegroJoystick^ joy)
	{
		auto* nativeJoy = joy->GetNativePointer();
		return al_get_joystick_num_buttons(nativeJoy);
	}

	Int32 Al::GetJoystickNumSticks(AllegroJoystick^ joy)
	{
		auto* nativeJoy = joy->GetNativePointer();
		return al_get_joystick_num_sticks(nativeJoy);
	}

	Int32 Al::GetNumJoysticks()
	{
		return al_get_num_joysticks();
	}

	JoyFlags Al::GetJoystickStickFlags(AllegroJoystick^ joy, Int32 stick)
	{
		auto* nativeJoy = joy->GetNativePointer();
		auto nativeFlags = al_get_joystick_stick_flags(nativeJoy, stick);
		return static_cast<JoyFlags>(nativeFlags);
	}

	String^ Al::GetJoystickAxisName(AllegroJoystick^ joy, Int32 stick, Int32 axis)
	{
		auto* nativeJoy = joy->GetNativePointer();
		const auto* nativeReturnValue = al_get_joystick_axis_name(nativeJoy, stick, axis);
		return gcnew String(nativeReturnValue);
	}

	String^ Al::GetJoystickButtonName(AllegroJoystick^ joy, Int32 button)
	{
		auto* nativeJoy = joy->GetNativePointer();
		const auto* nativeReturnValue = al_get_joystick_button_name(nativeJoy, button);
		return gcnew String(nativeReturnValue);
	}

	String^ Al::GetJoystickName(AllegroJoystick^ joy)
	{
		auto* nativeJoy = joy->GetNativePointer();
		const auto* nativeReturnValue = al_get_joystick_name(nativeJoy);
		return gcnew String(nativeReturnValue);
	}

	String^ Al::GetJoystickStickName(AllegroJoystick^ joy, Int32 stick)
	{
		auto* nativeJoy = joy->GetNativePointer();
		const auto* nativeReturnValue = al_get_joystick_stick_name(nativeJoy, stick);
		return gcnew String(nativeReturnValue);
	}

	void Al::GetJoystickState(AllegroJoystick^ joy, AllegroJoystickState^ retState)
	{
		auto* nativeJoy = joy->GetNativePointer();
		auto* nativeRetState = retState->GetNativePointer();
		al_get_joystick_state(nativeJoy, nativeRetState);
	}

	void Al::ReleaseJoystick(AllegroJoystick^ joy)
	{
		auto* nativeJoy = joy->GetNativePointer();
		al_release_joystick(nativeJoy);
	}

	void Al::UninstallJoystick()
	{
		al_uninstall_joystick();
	}
}

File: Source/AllegroDotNet/Al.Joystick.cs
using SubC.AllegroDotNet.Enums;
using SubC.AllegroDotNet.Models;
using SubC.AllegroDotNet.Native;

namespace SubC.AllegroDotNet;

/// <summary>
/// This static class contains the Allegro 5 library methods.
/// </summary>
public static partial class Al
{
  public static bool InstallJoystick()
  {
    return Interop.Core.AlInstallJoystick() != 0;
  }

  public static void UninstallJoystick()
  {
    Interop.Core.AlUninstallJoystick();
  }

  public static bool IsJoystickInstalled()
  {
    return Interop.Core.AlIsJoystickInstalled() != 0;
  }

  public static bool ReconfigureJoysticks()
  {
    return Interop.Core.AlReconfigureJoysticks() != 0;
  }

  public static int GetNumJoysticks()
  {
    return Interop.Core.AlGetNumJoysticks();
  }

  public static AllegroJoystick? GetJoystick(int num)
  {
    var pointer = Interop.Core.AlGetJoystick(num);
    return NativePointer.Create<AllegroJoystick>(pointer);
  }

  public static void ReleaseJoystick(AllegroJoystick? joystick)
  {
    Interop.Core.AlReleaseJoystick(NativePointer.Get(joystick));
  }

  public static bool GetJoystickActive(AllegroJoystick? joystick)
  {
    return Interop.Core.AlGetJoystickActive(NativePointer.Get(joystick)) != 0;
  }

  public static string? GetJoystickName(AllegroJoystick? joystick)
  {
    var pointer = Interop.Core.AlGetJoystickName(NativePointer.Get(joystick));
    return CStringAnsi.ToCSharpString(pointer);
  }

  public static string? GetJoystickStickName(AllegroJoystick? joystick, int stick)
  {
    var pointer = Interop.Core.AlGetJoystickStickName(NativePointer.Get(joystick), stick);
    return CStringAnsi.ToCSharpString(pointer);
  }

  public static string? GetJoystickAxisName(AllegroJoystick? joystick, int stick, int axis)
  {
    var pointer = Interop.Core.AlGetJoystickAxisName(NativePointer.Get(joystick), stick, axis);
    return CStringAnsi.ToCSharpString(pointer);
  }

  public static string? GetJoystickButtonName(AllegroJoystick? joystick, int button)
  {
    var pointer = Interop.Core.AlGetJoystickButtonName(NativePointer.Get(joystick), button);
    return CStringAnsi.ToCSharpString(pointer);
  }

  public static JoyFlags GetJoystickStickFlags(AllegroJoystick? joystick, int stick)
  {
    return (JoyFlags)Interop.Core.AlGetJoystickStickFlags(NativePointer.Get(joystick), stick);
  }

  public static int GetJoystickNumSticks(AllegroJoystick? joystick)
  {
    return Interop.Core.AlGetJoystickNumSticks(NativePointer.Get(joystick));
  }

  public static int GetJoystickNumAxes(AllegroJoystick? joystick, int stick)
  {
    return Interop.Core.AlGetJoystickNumAxes(NativePointer.Get(joystick), stick);
  }

  public static int GetJoystickNumButtons(AllegroJoystick? joystick)
  {
    return Interop.Core.AlGetJoystickNumButtons(NativePointer.Get(joystick));
  }

  public static void GetJoystickState(AllegroJoystick? joystick, ref AllegroJoystickState state)
  {
    Interop.Core.AlGetJoystickState(NativePointer.Get(joystick), ref state);
  }

  public static AllegroEventSource? GetJoystickEventSource()
  {
    var pointer = Interop.Core.AlGetJoystickEventSource();
    return NativePointer.Create<AllegroEventSource>(pointer);
  }
}
